player/lua/stats.lua

```diff
@@ -867,24 +867,90 @@ local function eval_ass_formatting()
+--   tabstops are 8, 16, 24..., and the output starts at 0 or a tabstop
+-- note: if maxwidth <= 2 and s doesn't fit: the result is "..." (more than 2)
+function term_ellipsis(s, maxwidth)
+    local TAB, ESC, SGR_END = 9, 27, ("m"):byte()
+    local width, ellipsis = 0, "..."
+    local fit_len, in_sgr
+
+    for i = 1, #s do
+        local x = s:byte(i)
+
+        if in_sgr then
+            in_sgr = x ~= SGR_END
+        elseif x == ESC then
+            in_sgr = true
+            ellipsis = "\27[0m..."  -- ensure SGR reset
+        elseif x < 128 or x >= 192 then  -- non UTF8-continuation
+            -- tab adds till the next stop, else add 1
+            width = width + (x == TAB and 8 - width % 8 or 1)
+
+            if fit_len == nil and width > maxwidth - 3 then
+                fit_len = i - 1  -- adding "..." still fits maxwidth
+            end
+            if width > maxwidth then
+                return s:sub(1, fit_len) .. ellipsis
+            end
+        end
+    end
+
+    return s
+end
+
+local function term_ellipsis_array(arr, from, to, maxwidth)
+    for i = from, to do
+        arr[i] = term_ellipsis(arr[i], maxwidth)
+    end
+    return arr
+end
+
+-- split str into a table
+-- example: local t = split(s, "\n")
+-- init: where to start looking for the separator pattern (default 1)
+-- plain: whether pat is a plain string (default false - pat is a pattern)
+local function split(str, pat, init, plain)
+    init = init or 1
+    local r, i, find, sub = {}, 1, string.find, string.sub
+    repeat
+        local f0, f1 = find(str, pat, init, plain)
+        r[i], i = sub(str, init, f0 and f0 - 1), i+1
+        init = f0 and f1 + 1
+    until f0 == nil
+    return r
+end
 
 -- Composes the output with header and scrollable content
 -- Returns string of the finished page and the actually chosen offset
 --
 -- header   : table of the header where each entry is one line
 -- content  : table of the content where each entry is one line
 -- offset   : the desired scroll offset of the content from the first line at the top
-local function compose_page(header, content, offset)
-    -- up to 22 lines for the terminal - so that mpv can also print
-    -- the status line without scrolling, and up to 40 lines for libass
-    -- because it can put a big performance toll on libass to process
-    -- many lines which end up outside (below) the screen.
-    local max_content_lines = (o.use_ass and 40 or 22) - #header
-    -- in the terminal the scrolling should stop once the last line is visible
-    local max_offset = o.use_ass and #content or #content - max_content_lines + 1
-    local from = max(1, min((offset or 1), max_offset))
-    local to = min(#content, from + max_content_lines - 1)
-    return table.concat(header) .. table.concat(content, "", from, to), from
+-- max_lines: (optional) limit number of output lines to this,
+--            values <= 0 or nil mean unlimited and scrolling is ignored
```

Comment:
If I'm reading it right, this comment is incorrect, because if `max_lines` is `nil` then it becomes 22 or 40.

Also, the word "scroll" is never mentioned elsewhere, so it's unclear what "scrolling is ignored" means. I think you mean "offset is considered 1" ? (why?)

Also, it looks to me like `compose_page` is always called with empty `header`, offset 1, and `nil` `max_lines`, which means:
- The output will be more than 22/40 elements if the header is not empty.
- The default page is limited to 22 terminal-lines / 40-array-elements eventhough it's not scrollable. That's a problem I think.
- Some of the argument might be redundant if they're unused and/or always used with the same value?

Reply:
> Also, it looks to me like `compose_page` is always called with empty `header`, offset 1, and `nil` `max_lines`

Apologies, my bad.

That's how it's called at the new places which this commit adds, which are the non-scrollable default page, and two places at teh non-scrollable cache page. amd it's fine, because they're not scrollable.

The other places where it's called is pages 0, 2, and 4, where it is called with useful parameters.

However, I still think we can consider simplifying the interface to `finalize_page(header, content, apply_scroll)`. Thoughts?

Reply:
We need max_width for bindlist, but max_lines can be replaced with a boolean.

Reply:
Not if we have an override option which bindlist can set.

Reply:
Yes, I should have tried actually implementing it before commenting, then I might have noticed.